### PR TITLE
Enable WC Pay Subscriptions during WC onboarding when US merchant selects subscriptions as product type

### LINKED ISSUE #6512: Enable WC Pay Subscriptions feature flag if chosen during WC onboarding
#### 🟢 Unblocked.

- [x] ~~https://github.com/Automattic/woocommerce-payments/issues/6511~~

As part of our move towards a single subscriptions offering, WC Pay subscriptions functionality should no longer be turned on by default for eligible stores (ref https://github.com/Automattic/woocommerce-payments/issues/6511).

However, there exists a flow in the WooCommerce onboarding where the user can select the Subscriptions product type and it will install WC Payments to get subscriptions functionality for free. 

<p align="center">
Screenshot 2023-06-13 at 3 29 21 pm
</p>

To make sure this flow is still honoured, we need to make sure that if a user selects this option during on-boarding, they still have WCPay Subscriptions enabled after installing WC Payments.

### Acceptance criteria

- [ ] If the free subscription option is chosen during WooCommerce onboarding, the feature should still remain enabled by default.

### Testing instructions

1. On a fresh site install WooCommerce
2. Follow the onboarding flow enter a US base address during the first step
3. Later in the onboarding flow, select Subscriptions as a product type (it should be free as long as you entered a US based address). 
4. Complete onboarding. 
5. WC Payments should be installed
   - With [#6511](https://github.com/Automattic/woocommerce-payments/issues/6511) subscriptions functionality won't be enabled. 

### Dev notes

As part of this project we will be removing this onboarding flow from WC. So, for this issue it will likely include a WC version check (eg wc version pre 8.0) + a check on the `woocommerce_onboarding_profile` option and perhaps a function hooked onto `woocommerce_onboarding_profile_data_updated` to determine if the store is eligible to have WCPay subscriptions enabled. 

### Additional context

Part of the Subscriptions Primary Offering project - peNR48-2u-p2

### PR DESCRIPTION
**Pull request by mattallan**
Fixes #6512

### Changes proposed in this Pull Request

In https://github.com/Automattic/woocommerce-payments/pull/6635 we changed the default behavior of our WCPay Subscriptions feature so that it is no longer enabled by default for all US-based stores.

The changes in this pull request make it so that when a US merchant goes through the WooCommerce onboarding wizard and selects the "free" subscriptions option (see screenshot below), the WCPay Subscriptions feature will be enabled.

image

> **Note** - The WooCommerce onboarding wizard is changing in the upcoming 7.9.0 release, so the changes in this PR are only for stores running versions < 7.9.0.

#### Implementation 
This pull request introduces a new function `maybe_enable_wcpay_subscriptions_after_onboarding()` that is hooked onto `woocommerce_onboarding_profile_data_updated` which is fired whenever the `woocommerce_onboarding_profile` wp-option is updated. On this hook, I'm checking if the `product_types` data is being updated and includes the `'subscriptions'` product type, if so, I verify the store is eligible for WC Pay Subscriptions before enabling the feature.

We only need this code to run on stores with WC versions < 7.9.0 so I've surrounded adding the `woocommerce_onboarding_profile_data_updated` hook with a version check.

### Testing instructions

There are a few key flows that these changes need to be tested on:

#### 1. Onboarding a US Store with WC <= 7.8.1

1. Spin up a fresh JN site with WooCommerce
2. Upload this branch and activate WC Payments
3. [Manually trigger the WC onboarding wizard](https://woocommerce.com/document/woocommerce-setup-wizard/#reusing-the-setup-wizard)
4. On the Product Types step make sure you select at least "Subscriptions"
image
5. After completing the WC Onboarding, confirm WCPay Subscriptions is enabled either by confirming the **WooCommerce > Subscriptions** menu item is loading.
6. Go into the wp-options table and delete the `_wcpay_feature_subscriptions` row
7. Confirm the Subscriptions menu item is no longer there
8. Go through the WC Onboarding Wizard again and don't select Subscriptions
9. Confirm WCPay Subscriptions feature is not enabled.

#### 2. Onboarding a non-US store running WC <= 7.8.1

1. Spin up a fresh JN site with WooCommerce
2. Go into **WooCommerce > Settings** and change your **Country / State** to Australia - Queensland
3. Upload this branch and activate WC Payments
4. [Manually trigger the WC onboarding wizard](https://woocommerce.com/document/woocommerce-setup-wizard/#reusing-the-setup-wizard)
5. On the Product Types step make sure you select at least "Subscriptions" (same screenshot above)
6. After completing the WC Onboarding, confirm WCPay Subscriptions is not enabled
7. Go into the wp-options table and delete the `_wcpay_feature_subscriptions` row

#### 3. Onboarding a store with WC > 7.9.0

1. Install and activate the latest [WC 7.9.0-beta-2](https://downloads.wordpress.org/plugin/woocommerce.7.9.0-beta.2.zip) zip
2. Install this branch of WC Payments
3. [Manually trigger the WC onboarding wizard](https://woocommerce.com/document/woocommerce-setup-wizard/#reusing-the-setup-wizard)
4. You'll notice the onboarding process looks a little different
image
5. Complete the new onboarding and make sure WCPay Subscriptions is not enabled.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6643` or branch name `issue/6512` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=issue/6512) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f9c4a53270394a9eb2733a5da83bd082e56a29a8
- Build time: 2023-07-13 00:03:58 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: changelog/issue-6512 [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: update
+
+Check WCPay Subscriptions eligibility after store completes WC onboarding wizardd.
```
Comment (james-allan) [inline changelog/issue-6512]:
```suggestion
Check WCPay Subscriptions eligibility after store completes WooCommerce onboarding wizard.
```

File: woocommerce-payments.php [hunk 2]
```diff
@@ -35,6 +35,8 @@
 function wcpay_activated() {
 	// Do not take any action if activated in a REST request (via wc-admin).
```
Comment (james-allan) [inline woocommerce-payments.php]:
```suggestion
	// When WooCommerce Payments is installed and activated from the WooCommerce onboarding wizard (via wc-admin REST request), check if the site is eligible for subscriptions.
```

This comment is no longer correct when it says "do not take any action" now that there is some action taken during that flow. So I've suggested we replaced it with a tweaked version of your new comment.

File: woocommerce-payments.php [hunk 3]
```diff
@@ -35,6 +35,8 @@
...
 	if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+		// When WooCommerce Payments is installed and activated from the WooCommerce onboarding wizard, check if the site is eligible for subscriptions.
```
Comment (james-allan) [inline woocommerce-payments.php]:
```suggestion
```

**Review by james-allan — CHANGES_REQUESTED** [review]
Changes look good. I still need to go through all the possible flows but I reached EOD so I'll leave these suggestions and retest tomorrow.

**Review by james-allan — APPROVED** [review]
Changes look good and work well for me. 

I had to hack it a bit to test the flow where the WC onboarding flow installs Woo Payments. Because I needed to test this version of the plugin, if this version of WC Payments is installed, then the WC onboarding flow won't activate the plugin.

Anyway, I hacked the WC onboarding flow to make sure it does activate WC Payments and confirmed that this code does enable the subscriptions feature.

**Pull request merged**